File: packages/sdk-typescript/src/resources/service-accounts.ts
import type { VoltxClient } from "../client.js";
import type { components } from "../generated/schema.js";
import type { CreateServiceAccountTokenResult, ServiceAccount, ServiceAccountToken } from "../types.js";

export type CreateServiceAccountInput = components["schemas"]["CreateServiceAccountDto"];
export type CreateServiceAccountTokenInput = components["schemas"]["CreateServiceAccountTokenDto"];

export class ServiceAccountsResource {
  constructor(private readonly client: VoltxClient) {}

  list(organizationId: string): Promise<ServiceAccount[]> {
    return this.client.get<ServiceAccount[]>(`/organizations/${organizationId}/service-accounts`);
  }

  create(organizationId: string, input: CreateServiceAccountInput): Promise<ServiceAccount> {
    return this.client.post<ServiceAccount>(`/organizations/${organizationId}/service-accounts`, input);
  }

  get(organizationId: string, id: string): Promise<ServiceAccount> {
    return this.client.get<ServiceAccount>(`/organizations/${organizationId}/service-accounts/${id}`);
  }

  suspend(organizationId: string, id: string): Promise<ServiceAccount> {
    return this.client.post<ServiceAccount>(
      `/organizations/${organizationId}/service-accounts/${id}/suspend`,
    );
  }

  reactivate(organizationId: string, id: string): Promise<ServiceAccount> {
    return this.client.post<ServiceAccount>(
      `/organizations/${organizationId}/service-accounts/${id}/reactivate`,
    );
  }

  listTokens(organizationId: string, id: string): Promise<ServiceAccountToken[]> {
    return this.client.get<ServiceAccountToken[]>(
      `/organizations/${organizationId}/service-accounts/${id}/tokens`,
    );
  }

  createToken(
    organizationId: string,
    id: string,
    input: CreateServiceAccountTokenInput,
  ): Promise<CreateServiceAccountTokenResult> {
    return this.client.post<CreateServiceAccountTokenResult>(
      `/organizations/${organizationId}/service-accounts/${id}/tokens`,
      input,
    );
  }

  revokeToken(organizationId: string, id: string, tokenId: string): Promise<void> {
    return this.client.delete<void>(
      `/organizations/${organizationId}/service-accounts/${id}/tokens/${tokenId}`,
    );
  }
}

File: packages/sdk-python/voltx_sdk/resources/service_accounts.py
from __future__ import annotations

from typing import TYPE_CHECKING, List

from ..generated.schema import CreateServiceAccountDto, CreateServiceAccountTokenDto
from ..types import CreateServiceAccountTokenResult, ServiceAccount, ServiceAccountToken

if TYPE_CHECKING:
    from ..client import VoltxClient


class ServiceAccountsResource:
    def __init__(self, client: "VoltxClient") -> None:
        self._client = client

    def list(self, organization_id: str) -> List[ServiceAccount]:
        return self._client.get(f"/organizations/{organization_id}/service-accounts")

    def create(self, organization_id: str, input: CreateServiceAccountDto) -> ServiceAccount:
        return self._client.post(f"/organizations/{organization_id}/service-accounts", input)

    def get(self, organization_id: str, id: str) -> ServiceAccount:
        return self._client.get(f"/organizations/{organization_id}/service-accounts/{id}")

    def suspend(self, organization_id: str, id: str) -> ServiceAccount:
        return self._client.post(f"/organizations/{organization_id}/service-accounts/{id}/suspend")

    def reactivate(self, organization_id: str, id: str) -> ServiceAccount:
        return self._client.post(f"/organizations/{organization_id}/service-accounts/{id}/reactivate")

    def list_tokens(self, organization_id: str, id: str) -> List[ServiceAccountToken]:
        return self._client.get(f"/organizations/{organization_id}/service-accounts/{id}/tokens")

    def create_token(
        self, organization_id: str, id: str, input: CreateServiceAccountTokenDto
    ) -> CreateServiceAccountTokenResult:
        return self._client.post(f"/organizations/{organization_id}/service-accounts/{id}/tokens", input)

    def revoke_token(self, organization_id: str, id: str, token_id: str) -> None:
        self._client.delete(f"/organizations/{organization_id}/service-accounts/{id}/tokens/{token_id}")
